File: UTCS/UTCS/Classes/Views/UTCSLabView.m
/*
//
//  UTCSLabView.m
//  UTCS
//
//  Created by Comyar Zaheri on 4/25/14.
//  Copyright (c) 2014 UTCS. All rights reserved.
//

#import "UTCSLabView.h"
#import "UTCSLabMachineView.h"
#import "UTCSLabViewLayout.h"
#import "UTCSLabViewLayoutAttributes.h"

@interface UTCSLabView ()
@property (nonatomic) NSArray *machineViews;
@end


@implementation UTCSLabView

- (instancetype)initWithFrame:(CGRect)frame layout:(UTCSLabViewLayout *)layout
{
    if (self = [super initWithFrame:frame]) {
        _layout = layout;
        [self prepareLayout];
    }
    return self;
}

- (void)invalidateLayout
{
    [self prepareLayout];
}

- (void)prepareLayout
{
    [_layout prepareLayoutForLabView:self];
    
    for (UTCSLabMachineView *machineView in self.machineViews) {
        [machineView removeFromSuperview];
    }
    self.machineViews = nil;
    
    NSMutableArray *machineViews = [NSMutableArray new];
    
    for (int i = 0; i < _layout.numberOfLabMachines; ++i) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:i inSection:0];
        
        UTCSLabMachineView *labMachineView = [UTCSLabMachineView new];
        labMachineView.tag = indexPath.row;
        
        [machineViews addObject:labMachineView];
    }
    
    self.machineViews = machineViews;
}

- (UTCSLabMachineView *)dequeueMachineViewForIndexPath:(NSIndexPath *)indexPath
{
    return self.machineViews[indexPath.row];
}

- (void)reloadData
{
    NSInteger count = [self.machineViews count];
    
    for (int i = 0; i < count; ++i) {
        NSIndexPath *indexPath = [NSIndexPath indexPathForItem:i inSection:0];
        NSString *machineName = [_layout labMachineNameForIndexPath:indexPath];
        UTCSLabViewLayoutAttributes *layoutAttributes = [_layout layoutAttributesForIndexPath:indexPath];
        UTCSLabMachineView *machineView = [self.dataSource labView:self machineViewForIndexPath:indexPath name:machineName];
        
        machineView.frame = CGRectMake(0.0, 0.0, layoutAttributes.size.width, layoutAttributes.size.height);
        machineView.center = layoutAttributes.center;
        machineView.tag = indexPath.row;
        
        [self addSubview:machineView];
    }
}

@end

*/

File: UTCS/UTCS/Classes/UTCSLabView.swift
//
//  UTCSLabView.swift
//  UTCS
//
//  Created by Taylor Schmidt on 4/8/16.
//  Copyright © 2016 UTCS. All rights reserved.
//
import UIKit

import Foundation

protocol UTCSLabViewDataSource {
    func labView(labView: UTCSLabView, machineViewForIndexPath: NSIndexPath, name: String) -> UTCSLabMachineView
}

class UTCSLabView : UIView {
    var machineViews = [UTCSLabMachineView]()
    var layout: UTCSLabViewLayout
    var dataSource: UTCSLabViewDataSource?
    init(frame: CGRect, layout: UTCSLabViewLayout) {
        self.layout = layout
        super.init(frame: frame)
        self.prepareLayout()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func invalidateLayout() {
        self.prepareLayout()
    }
    
    func prepareLayout() {
        self.layout.prepareLayoutForLabView(self)
        for machineView in self.machineViews { machineView.removeFromSuperview() }
        self.machineViews = []
        
        for i in 0..<self.layout.numberOfLabMachines {
            let indexPath = NSIndexPath(forRow: i, inSection: 0)
            let labMachineView = UTCSLabMachineView()
            labMachineView.tag = indexPath.row
            self.machineViews.append(labMachineView)
        }
    }
    
    func dequeueMachineViewForIndexPath(indexPath: NSIndexPath) -> UTCSLabMachineView {
        return self.machineViews[indexPath.row]
    }
    
    func reloadData() {
        let count = self.machineViews.count
        for i in 0..<count {
            let indexPath = NSIndexPath(forItem: i, inSection: 0) // should this be forRow?
            let machineName = self.layout.labMachineNameForIndexPath(indexPath)
            guard let layoutAttributes = self.layout.layoutAttributesForIndexPath(indexPath) else {
                // some error code?
                return
            }
            let machineView = self.dataSource!.labView(self, machineViewForIndexPath: indexPath, name: machineName!)
            machineView.frame = CGRectMake(0.0, 0.0, layoutAttributes.size.width, layoutAttributes.size.height)
            machineView.center = layoutAttributes.center
            machineView.tag = indexPath.row
            self.addSubview(machineView)
        }
    }
    
}